ruby/ql/lib/codeql/ruby/ast/Control.qll

```diff
@@ -422,6 +422,189 @@ class WhenExpr extends Expr, TWhenExpr {
   }
 }
 
+/**
+ * A `case` statement used for pattern matching.
+ */
+class CaseMatch extends ControlExpr, TCaseMatch {
+  private Ruby::CaseMatch g;
+
+  CaseMatch() { this = TCaseMatch(g) }
+
+  final override string getAPrimaryQlClass() { result = "CaseMatch" }
+
+  /**
+   * Gets the expression being compared. For example, `foo` in the following example.
+   * ```rb
+   * case foo
+   * in 0
+   *   puts 'zero'
+   * in 1
+   *   puts 'one'
+   * end
+   * ```
+   */
+  final Expr getValue() { toGenerated(result) = g.getValue() }
+
+  /**
+   * Gets the `n`th branch of this case expression, either an `InClause` or a
+   * `StmtSequence`.
+   */
+  final Expr getBranch(int n) {
+    toGenerated(result) = g.getClauses(n)
+    or
+    n = count(g.getClauses(_)) and toGenerated(result) = g.getElse()
+  }
+
+  /**
+   * Gets a branch of this case expression, either an `InClause` or an
+   * `StmtSequence`.
+   */
+  final Expr getABranch() { result = this.getBranch(_) }
+
+  /** Gets a `in` clause of this case expression. */
+  final InClause getInClause(int n) { result = this.getBranch(n) }
+
+  /** Gets the `n`th `in` clause of this case expression. */
+  final InClause getAnInClause() { result = this.getABranch() }
+
+  /** Gets the `else` branch of this case expression, if any. */
+  final StmtSequence getElseBranch() { result = this.getABranch() }
+
+  /**
+   * Gets the number of branches of this case expression.
+   */
+  final int getNumberOfBranches() { result = count(this.getBranch(_)) }
+
+  final override string toString() { result = "case ... in" }
+
+  override AstNode getAChild(string pred) {
+    result = super.getAChild(pred)
+    or
+    pred = "getValue" and result = this.getValue()
+    or
+    pred = "getBranch" and result = this.getBranch(_)
+  }
+}
+
+/**
+ * An `in` clause of a `case` expression.
+ * ```rb
+ * case foo
+ * in [ a ] then a
+ * end
+ * ```
+ */
+class InClause extends Expr, TInClause {
+  private Ruby::InClause g;
+
+  InClause() { this = TInClause(g) }
+
+  final override string getAPrimaryQlClass() { result = "InClause" }
+
+  /** Gets the body of this case-in expression. */
+  final Stmt getBody() { toGenerated(result) = g.getBody() }
+
+  /**
+   * Gets the pattern in this case-in expression. In the
+   * following example, the pattern is `Point{ x:, y: }`.
+   * ```rb
+   * case foo
+   * in Point{ x:, y: }
+   *   x + y
+   * end
+   * ```
+   */
+  final CasePattern getPattern() { toGenerated(result) = g.getPattern() }
+
+  /**
+   * Gets the pattern guard in this case-in expression. In the
+   * following example, the pattern guard is `x > 10`.
+   * ```rb
+   * case foo
+   * in [ x ] if x > 10 then ...
+   * end
+   * ```
+   */
+  final PatternGuard getPatternGuard() { toGenerated(result) = g.getGuard() }
```

Comment:
Make sense except that we would end-up writing `getCondition().getCondition()` to the the `expr` part of `if expr` and `unless expr`. This would look a bit silly. However, I think we should be able to get rid of the intermediate node.

Reply:
Sounds good to get rid of `PatternGuard` altogether. We will then need a predicate along the lines of `InClause::conditionIsUnless()`.

Reply:
Done! I added `hasIfCondition` and `hasUnlessCondition` for symmetry.